### PR TITLE
refactor: make errors more informative

### LINKED ISSUE #210: chore: make errors more informative
The error classes are messy.
A duplicate index error (`E11000` in mongo) should be one of the two:
- `FileExistsWithSameName` (name-parent-owner) 
- `FileExistsInS3` (bucket-key)

Both implements `UniqueIndexExistsError` which implements `ClientError`.
Both should

### PR DESCRIPTION
**Pull request by yonatandt**
resolves #210

### CODE REVIEW

File: src/file/file.service.ts [hunk 1]
```diff
@@ -158,18 +159,26 @@ export class FileService {
...
+    if (fileID === parentID) {
+      throw new ClientError(
+        status.INVALID_ARGUMENT,
+        `cyclic nesting error: Trying to put the file ${fileID} inside itself`);
+    }
     if (file.type === FolderContentType && await this.isFolderAnAncestor(parentID, fileID)) {
-      throw new ClientError('cyclic nesting error');
+      throw new ClientError(
+        status.FAILED_PRECONDITION,
+        `cyclic nesting error: The folder ${fileID} is an ancestor of ${parentID}, therefore it cant be inside it`);
```
Comment (Shahar-Y) [inline src/file/file.service.ts:173]: nice

File: src/utils/errors/application.error.ts [hunk 2]
```diff
@@ -1,32 +1,60 @@
...
+ * three pieces of data: error code, error message, and error details.
+ *
+ * You can find out more about this error model and how to work with it in the
+ * [API Design Guide](https://cloud.google.com/apis/design/errors).
+ *
+ * Unfortunately, the node's grpc package implements grpc.ServiceError
+ * differently from the Google's status code, as can be seen in the Interface.
+ * Therefore the client should pay attention to deserialize the status object
+ * of this service correctly.
+*/
```
Comment (Shahar-Y) [inline src/utils/errors/application.error.ts:20]: Good explanation

File: src/utils/mongo.error.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { MongoError } from 'mongodb';
+
+/**
+ * Extracts the indixes names and values thrown by the duplicate key error.
+ * @param error - the mongo error thrown.
+ * @return string with the index names and values.
+ */
+export function getMongoErrorIndixes(error: MongoError): { indexName: string, values: string } {
```
Comment (Shahar-Y) [inline src/utils/mongo.error.ts:8]: Indices* ? Indexes* ?

File: src/utils/mongo.error.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  // extract the indixes names in the MongoError
+  const indicesRegex: RegExp = new RegExp(/index\:\ (?:.*\.)?\$?(?:([_a-z0-9]*)(?:_\d*)|([_a-z0-9]*))\s*dup key/i);
+  const indicesMatch: RegExpMatchArray = error.message.match(indicesRegex);
+  let indexName: string = indicesMatch[1] || indicesMatch[2];
+
+  // prettify indixes names
+  const re: RegExp = new RegExp('_1_', 'g');
```
Comment (Shahar-Y) [inline src/utils/mongo.error.ts:15]: Are you sure the regex is good?

**Review by Shahar-Y — DISMISSED** [review]

File: src/utils/mongo.error.ts (hunk 3 above)
Comment (yonatandt) [reply src/utils/mongo.error.ts]: fixed

File: CHANGELOG.md [hunk 5]
```diff
@@ -9,9 +9,17 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Refactor
 
-- REFACTOR([95](https://github.com/meateam/drive-project/issues/96)): upgrade docker compose to v3 and use env_file
 - errors are more informative.
 
+## [v3.0.0] - 2020-11-16
+
+## Added
```
Comment (Shahar-Y) [inline CHANGELOG.md:16]: ###

File: src/utils/errors/client.error.ts [hunk 6]
```diff
@@ -1,62 +1,117 @@
...
+  /**
+   * The constructor of ArgumentInvalidError.
+   * @param argumentName The problematic argument's name.
+   * @param argumentValue The problematic argument value.
+   * @param message An optional custom message.
+   */
+  constructor(argumentName: string, argumentValue: any, message?: string) {
+    const metadata = new grpc.Metadata();
+    metadata.add('fieldName', argumentName);
+    metadata.add('filedValue', `${argumentValue}`);
```
Comment (Shahar-Y) [inline src/utils/errors/client.error.ts]: stringify

File: CHANGELOG.md [hunk 7]
```diff
@@ -5,6 +5,22 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
+
+### Refactor
+
+- errors are more informative.
```
Comment (Shahar-Y) [inline CHANGELOG.md:12]: add link to the PR

File: src/utils/logger.ts [hunk 8]
```diff
@@ -78,13 +77,13 @@ export function wrapper(func: Function) :
...
 
       const res = await func(call, callback);
-      apm.endTransaction(statusToString(grpc.status.OK));
+      apm.endTransaction(grpc.status[grpc.status.OK]);
       const resInfo: object = extractResLog(res);
       log(Severity.INFO, 'response', func.name, traceID, resInfo);
       callback(null, res);
     } catch (err) {
       const validatedErr : ApplicationError = validateGrpcError(err);
-      log(Severity.ERROR, func.name, err.message, getCurrTraceId());
```
Comment (Shahar-Y) [inline src/utils/logger.ts:87]: if you send `any` type to string, at least stringify it so it won't put [object objecy]

**Review by Shahar-Y — CHANGES_REQUESTED** [review]

File: src/utils/logger.ts (hunk 8 above)
Comment (yonatandt) [reply src/utils/logger.ts:87]: done

File: CHANGELOG.md (hunk 7 above)
Comment (yonatandt) [reply CHANGELOG.md]: done

File: src/utils/errors/client.error.ts (hunk 6 above)
Comment (yonatandt) [reply src/utils/errors/client.error.ts]: done

File: CHANGELOG.md (hunk 5 above)
Comment (yonatandt) [reply CHANGELOG.md]: done

**Review by Shahar-Y — DISMISSED** [review]

File: CHANGELOG.md [hunk 9]
```diff
@@ -5,6 +5,22 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
+
+### Refactor
+
+- errors are more informative [226].
```
Comment (Shahar-Y) [inline CHANGELOG.md:12]:
```suggestion
- errors are more informative ([226](https://github.com/meateam/file-service/pull/226)).
```

**Review by Shahar-Y — APPROVED** [review]

**Pull request closed without merging**